Geometry integration

aymanhab opened the pull request:
This PR merges work done to redefine what geometric shapes can be attached to OpenSim ModelComponents. The available new Geometry is written as  a set of Components that know how to draw themselves. Supported Geometries are: Mesh, Sphere, Brick, Ellipsoid, Cylinder, Cone, Line, Frame and Arrow. Geometry gets translated into SimTK::DecorativeGeometry objects that are sent to the Visualizer. With this PR, ModelComponents can display themselves either by implementing generateDecorations directly or by including Geometry components as (sub)components.

sherm1 commented:
I see you've been busy!

The convention for work-in-progress PRs is to put [WIP] at the beginning of the title -- that way everyone knows not to review yet. I fixed it.

aymanhab commented:
Thanks Sherm.

On Fri, Apr 24, 2015 at 11:31 AM, Michael Sherman <notifications@github.com>
wrote:

> I see you've been busy!
>
> The convention for work-in-progress PRs is to put [WIP] at the beginning
> of the title -- that way everyone knows not to review yet. I fixed it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opensim-org/opensim-core/pull/417#issuecomment-96026533>
> .
>

aymanhab commented:
@chrisdembia Is there a way to disable caching on AppVoyer or verify which version of Simbody is being used? The reported error is in code using a SimTK_Array which has been refactored lately for C++11 compliance. Thanks.

sherm1 commented:
I'm not seeing that now, Ayman. The AppVeyor errors [here](https://ci.appveyor.com/project/opensim-org/opensim-core/build/325) are:
```
1/64 Test #35: testExampleMainRuns ..........................***Failed    0.52 sec
ERROR- Ground::addComponent() 'ground' already has 'ground_geom_3' as a subcomponent.

37/64 Test #36: testExampleMain ..............................***Failed    0.72 sec
 
Exception:
  Storage: ERROR- failed to open file tugOfWar_states.sto
  file= C:\projects\opensim-core\OpenSim\Common\Storage.cpp
  line= 126
 
 
IO.OpenInputFile(const string&,openmode mode): failed to open tugOfWar_states.sto
```

aymanhab commented:
@sherm1 Both errors are downstream effects of a search in a SimTK::Array that reports duplicates on AppVoyer but runs fine locally and on travis. Neither of these occurred in a previous build on AppVoyer on 4/22 
https://ci.appveyor.com/project/opensim-org/opensim-core/build/313
No material code change other than Simbody's and the change of tabs into spaces!

sherm1 commented:
Oh! Was your local run using an earlier Simbody?

chrisdembia commented:
> @chrisdembia Is there a way to disable caching on AppVoyer or verify which version of Simbody is being used?

This should be the simbody commit that Appveyor is using: https://github.com/simbody/simbody/commit/155047a5fe4de79164a93825a8a005cf80f20b3a

sherm1 commented:
Uh oh. I've got this built now on my local machine. I see the problem in Release mode but it goes away in Debug!

sherm1 commented on OpenSim/Simulation/Model/PhysicalFrame.cpp:
```diff
@@ -45,11 +45,16 @@ PhysicalFrame::PhysicalFrame() : Frame()
 {
     setAuthors("Matt DeMers, Ayman Habib, Ajay Seth");
     constructProperties();
+    FrameGeometry* frm = new FrameGeometry(0.2);
```
This FrameGeometry object gets leaked at the end of this method. Since it is *copied* to the GeometrySet property, it shouldn't be heap allocated here.

sherm1 commented on OpenSim/Simulation/Model/ModelComponent.cpp:
```diff
@@ -76,7 +79,16 @@ void ModelComponent::extendConnect(Component &root)
+    int geomSize = getProperty_GeometrySet().size();
+    if (geomSize > 0){
+        for (int i = 0; i < geomSize; ++i){
+            addComponent(&upd_GeometrySet(i));
```
@aseth1 is this an OK use of `addComponent()`? It is in `finalizeFromProperties()` so gets called from the constructor.

aseth1 replied:
@sherm1 `MoelComponent::finalizeFromProperties()` is not being called from the constructor nor should it be. This is the correct place to be doing `addComponent()` as well.

sherm1 replied:
@aseth1, I'm confused. I'm sitting in the debugger with a breakpoint in this routine and can see that it is called from the constructor `Model::Model()`. [Here](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Simulation/Model/Model.cpp#L98) is the code (from the master branch). Please clarify.

aseth1 replied:
Sorry @sherm1 I thought you meant ModelComponent's constructor. I don't particularly like that Model construction is calling `finalizeFromProperties` either, but not because it is necessarily dangerous. We want to minimize calls to finalizeFromProperties because it could be loading big data files or doing some computation/initialization of modeling parameters based on property values, so we don't want deserialization or default construction to invoke it. The faux pas that should not be made is to call `addComponent()` outside of `finalizeFromProperties`.

sherm1 commented on OpenSim/Simulation/Model/ModelComponent.cpp:
```diff
@@ -84,19 +96,38 @@ void ModelComponent::connectToModel(Model& model)
+        
     }
+    append_GeometrySet(*geom);
+    addComponent(geom); 
```
Ayman, I don't think this can be doing what you meant it to do. `append_GeometrySet()` takes a `Geometry` object (not a pointer) and pushes a copy of that object onto the property list. Then `addComponent()` takes a pointer and stores it, with no ownership transfer. So (1) the component is not the same object as is in the `GeometrySet` property, and (2) the heap-allocated object `geom` which is supposedly getting adopted ends up getting leaked instead since no one owns it. Also, any operations that are subsequently performed on the subcomponent don't affect the object stored in the GeometrySet property.

sherm1 commented on OpenSim/Simulation/Model/ModelComponent.h:
```diff
@@ -79,9 +78,18 @@ class ModelDisplayHints;
+These are the serializable properties associated with this class. **/
+/**@{**/
+public:
+    OpenSim_DECLARE_LIST_PROPERTY(GeometrySet, Geometry,
```
I am confused by this property name. I thought it was an object but was unable to locate a `GeometrySet` class -- did I miss it? If it is not the name of an object, then it doesn't follow our convention for property names, which would be something like `geometry_set`.

sherm1 commented:
I'm still trying to nail this precisely, but here's what I've found so far in case it rings a bell. Geometry is being added to Ground() and to the `_components` array (an array of pointers that are references, not owners) via addComponent(). But there is already an entry in `_components`, in the first element, for a `FrameGeometry` object. That first entry (a pointer) is pointing to a garbage element that has already been destructed. Under some rare circumstances, that same heap address gets re-used for one of the later pieces of geometry, and `addComponent()` has a test for that which is where the error message is coming from. But the real problem is the `_components` entry referring to a destructed object.

This doesn't currently appear to have anything to do with the Array_ class mods except that it affects how many heap objects get allocated, and thus the chance that the deleted object will get re-used in time for addComponents() to notice it. I will continue to track it down.

aymanhab commented:
Sherm, thanks for the great investigative work while this is still in
progress. Please don't spend more time on it until I take another look. My
next commit will run Valgrind and that should shine light on this and any
other  memory problems. Cheers.
On Apr 27, 2015 5:34 PM, "Michael Sherman" <notifications@github.com> wrote:

> I'm still trying to nail this precisely, but here's what I've found so far
> in case it rings a bell. Geometry is being added to Ground() and to the
> _components array (an array of pointers that are references, not owners)
> via addComponent(). But there is already an entry in _components, in the
> first element, for a FrameGeometry object. That first entry (a pointer)
> is pointing to a garbage element that has already been destructed. Under
> some rare circumstances, that same heap address gets re-used for one of the
> later pieces of geometry, and addComponent() has a test for that which is
> where the error message is coming from. But the real problem is the
> _components entry referring to a destructed object.
>
> This doesn't currently appear to have anything to do with the Array_ class
> mods except that it affects how many heap objects get allocated, and thus
> the chance that the deleted object will get re-used in time for
> addComponents() to notice it. I will continue to track it down.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opensim-org/opensim-core/pull/417#issuecomment-96857805>
> .
>

sherm1 commented:
No need for valgrind, Ayman. The problem turned out to be simple: there is a property list of Geometry objects; the _component list points into that property list; when more geometry gets added the Geometry list gets reallocated (using standard `push_back()` behavior). That causes the objects to get copied; now they have new addresses and the `_component` list contains garbage.

That's the problem; the solution is much less obvious! I think we'll need to discuss this with @aseth1.

sherm1 commented:
This code shows [the problem](https://github.com/opensim-org/opensim-core/pull/417/files#diff-9d34dca86ce3c6c34a7e9c2dca87d7b4R88):
```c++
        for (int i = 0; i < geomSize; ++i){
            addComponent(&upd_GeometrySet(i));
```
The subcomponent list is given a pointer directly to an object that is stored in a list property. Later additions to that list cause reallocation, moving the objects around and destructing the old ones. But no one tells the subcomponent list about it.

If components are allowed to point into property lists, I would think the component list must be rebuilt after any property change has been made. Is there a mechanism for enforcing that? You can see [where it breaks](https://github.com/opensim-org/opensim-core/pull/417/files#diff-9d34dca86ce3c6c34a7e9c2dca87d7b4R126) in this PR:
```c++
void ModelComponent::adoptGeometry(OpenSim::Geometry* geom) {
    ...
    append_GeometrySet(*geom); // <--- reallocation occurs here, invalidating _component list
    addComponent(geom); // <--- ignore this; it neither fixes the broken list nor adds the right object to the end!
}
```

aseth1 commented:
>The subcomponent list is given a pointer directly to an object that is stored in a list property. Later additions to that list cause reallocation, moving the objects around and destructing the old ones. But no one tells the subcomponent list about it.
If components are allowed to point into property lists, I would think the component list must be rebuilt after any property change has been made. Is there a mechanism for enforcing that? You can see where it breaks in this PR:

The subcomponent list [IS](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Common/Component.cpp#L150) cleared and rebuilt when `finalizeFromProperties()` is called on a Component. It's `extendFinalizeFromProperties()` should add to the subcomponent list, which it is doing (especially if it has components that are held in properties!). I think the problem is that the adopting of geometry is effectively editing the properties and it should not simultaneously be marking these geometry as subcomponents.

sherm1 commented:
Thanks for the clarifications, Ajay -- makes sense. So then finalizeFromProperties() is called once from the `Model()` constructor and then again in `initSystem()` after all the properties have been modified. And that second call rebuilds the `_components` list from scratch?

sherm1 commented:
Ayman, a general comment: the various leaks and heap trashings can be traced directly to excessive use of explicit heap management. Almost every place there was a `new` there was a leak! And where `addDecorativeGeometry(object)` got changed to `adoptGeometry(pointer)` there were multiple problems including leaks, trashing, and putting the wrong object on lists. That is typical for C++ code written this way and is why you should avoid it! Ideally: no pointers, no news, no deletes, no `&`. Where they can't be avoided (rare), expect problems.

aymanhab commented:
Thanks  Sherm for the advice. I think the issue at the core of this is my
trying to use Simtk Array which makes internal copies of objects and
manages them the modern way the same way as Opensim Sets which are used
everywhere else and they indeed require using new, heap - allocation and
all the things we'd like to avoid. I'd love to work with you to make this
work as intended. I did run Valgrind yesterday and will be looking at its
output to see what other issues or places to fix.
On Apr 28, 2015 9:23 AM, "Michael Sherman" <notifications@github.com> wrote:

> Ayman, a general comment: the various leaks and heap trashings can be
> traced directly to excessive use of explicit heap management. Almost every
> place there was a new there was a leak! And where
> addDecorativeGeometry(object) got changed to adoptGeometry(pointer) there
> were multiple problems including leaks, trashing, and putting the wrong
> object on lists. That is typical for C++ code written this way and is why
> you should avoid it! Ideally: no pointers, no news, no deletes, no &.
> Where they can't be avoided (rare), expect problems.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opensim-org/opensim-core/pull/417#issuecomment-97125852>
> .
>

sherm1 commented:
I'd be happy to work on this with you, Ayman. I don't think it is anything major, just a matter of changing memory management style to follow a safer discipline.

aseth1 commented:
>Thanks for the clarifications, Ajay -- makes sense. So then finalizeFromProperties() is called once from the Model() constructor and then again in initSystem() after all the properties have been modified. And that second call rebuilds the _components list from scratch?

Yes. Anytime `finalizeFromProperties()` is called the subcomponent list should be rebuilt from scratch. That was the only way I could guarantee consistency and seemed to add little overhead. I think we can remove the first call in the constructor. Way back when Model used to call `setup()` but setup was decomposed into more distinct operations (finalizing from props, wiring up components (connect) , ...) . The call to `finalizeFromProperties()` in the constructor may be purely legacy in which case I will investigate removing it. @aymanhab any reason to keep it?

aymanhab commented:
@aseth1 I have no reason to keep it, it's not used from the GUI as far as I can tell (we may want to go back and find out why it was added in the first place).

aymanhab commented:
@sherm1 I believe this's now reviewable, should I remove the WIP annotation or create a fresh one?

sherm1 commented:
> should I remove the WIP annotation or create a fresh one?

Just edit the title to get rid of the [WIP] and ask for reviews.

aymanhab commented:
Thanks @sherm1, I found a bug in model conversion that I'll fix first then will change the label.

aymanhab commented:
@sherm1 @aseth1 @chrisdembia @tkuchida This is functional to the extent that I can take an old model and visualize it in a build of the GUI. Documentation, examples and test cases included. Some loose ends are parts of surfaces and handling Wrapping/Contact as well as texture will be handled downstream.

sherm1 commented on OpenSim/Common/ComponentList.h:
```diff
@@ -168,6 +168,10 @@ class ComponentListIterator :
         return _node == &*iter;
     }
+    /// Check for equality using method rather than operator to support scripting
+    bool equals(const ComponentListIterator& iter) const {
```
Shouldn't this method have the exact same implementation as `operator==`? It seems very different -- I think one is comparing node addresses and the other is comparing iterators. I would like to see just one implementation (the right one, whichever that is!), for `equals()`, and then `operator==` should just invoke that implementation.

aymanhab replied:
@sherm1 My understanding of the syntax is that the method is just invoking the operator (the operator came first), please let me know if that's not the case or suggest the correct syntax. Both have been used and behave as expected but that maybe just a coincidence.

sherm1 replied:
> My understanding of the syntax is that the method is just invoking the operator

Oh, I see. Yes, that's probably right but very confusing and not at all obvious because there are so many potential overloads for `==`. To make it clear, the `equals()` method should contain the real implementation, the `operator==` should just invoke it.

sherm1 commented on OpenSim/Examples/CustomActuatorExample/toyLeg_example.cpp:
```diff
@@ -65,33 +65,35 @@ int main()
+        cyl.set_scale_factors(linkageDimensions);
+        Frame* cyl1Frame = new PhysicalOffsetFrame(*linkage1, Transform(Vec3(0.0, linkageLength / 2.0, 0.0)));
+        cyl1Frame->setName("Cyl1_frame");
+        osimModel.addFrame(cyl1Frame);
```
Shouldn't this be `osimModel.adoptFrame()`?

aymanhab replied:
Agree it should but that would be inconsistent with all other model components that get "adopted" using "add" e.g. addBody, addJoint, ... etc. accordingly there's no adoptFrame method.

sherm1 replied:
> would be inconsistent with all other model components that get "adopted" using "add"

Oh :cry:. @aseth1 should these Model methods have their names changed to "adopt"? I did that in Simbody recently in a few places and left the "add" methods around as deprecated so old code would still work. Maybe file an issue?

aseth1 replied:
While I agree that `adopt` is more accurate and I am happy that our containers have the adopt and append semantics, when it comes to building a model programmatically, I think it is less intuitive to user developers and scripting users than `addWhatever()`. I think you should file an issue so we can discuss it. 

If we had a generic `Component::add<ComponentType>(ComponentType& comp)` it would eliminate the need for individual add methods.  We could eliminate Sets by instead enabling Component to build/curate its own internal lists based on whatever type is being added. It would still be fast to find and access specific components via a `get/upd<ComponentType>(name)` capability. This seems both more compact and scalable than having `addBody`, `addJoint`, `addConstraint`, `addForce`, `addProbe`, `addFrame`, `addGeometry` which is unsustainable!

I am also not a fan of the current `addWhatever(Whatever* thingy)` form which takes a pointer to newed Whatever. We could reference count (perhaps via a ComponentHandle) so that Components being added can always be allocated on the stack (like with MobilizedBody) by the user.

sherm1 commented on OpenSim/Simulation/Model/AppearanceMap.h:
```diff
@@ -0,0 +1,100 @@
+ * National Institutes of Health (U54 GM072970, R24 HD065690) and by DARPA    *
+ * through the Warrior Web program.                                           *
+ *                                                                            *
+ * Copyright (c) 2005-2014 Stanford University and the Authors                *
```
2015?

sherm1 commented on OpenSim/Simulation/Model/AppearanceMap.h:
```diff
@@ -0,0 +1,100 @@
+
+// INCLUDE
+#include <iostream>
+#include <math.h>
```
This is not right for C++: should be `#include <cmath>` which introduces the same symbols in the `std::` namespace (and also in the global namespace).

sherm1 replied:
I'm not sure if we have a standard for this, but I suggest always putting the system header files *after* the local ones rather than before. That makes sure that the local ones are self contained.

sherm1 commented on OpenSim/Simulation/Model/ExpressionBasedBushingForce.cpp:
```diff
@@ -609,7 +605,7 @@ void ExpressionBasedBushingForce::updateDisplayer(const SimTK::State& s)
     _model->getSimbodyEngine().transformPosition(s, body2, get_location_body_2(),
                                                  globalLocation2);
-
+    /*
```
why not remove this code?

aymanhab replied:
@sherm1 You're right I'll remove all these. I kept some around in development to reuse/replicate in generateDecorations when it wasn't implemented but it's not worth it with the added clutter.

sherm1 commented on OpenSim/Simulation/Model/ExpressionBasedBushingForce.cpp:
```diff
@@ -617,15 +613,17 @@ void ExpressionBasedBushingForce::updateDisplayer(const SimTK::State& s)
 
 void ExpressionBasedBushingForce::updateGeometry(const SimTK::State& s)
 {
+    /*
```
and this

sherm1 commented on OpenSim/Simulation/Model/ExpressionBasedPointToPointForce.cpp:
```diff
@@ -123,22 +118,22 @@ void ExpressionBasedPointToPointForce::updateDisplayer(const SimTK::State& s)
                                                  globalLocation2);
 
-    if (_displayer.countGeometry()==0){
+/*  if (_displayer.countGeometry()==0){
```
remove? (and below)

sherm1 commented on OpenSim/Simulation/Model/FunctionBasedBushingForce.cpp:
```diff
@@ -509,23 +506,23 @@ void FunctionBasedBushingForce::updateDisplayer(const SimTK::State& s)
     _model->getSimbodyEngine().transformPosition(s, body2, get_location_body_2(),
                                                  globalLocation2);
-
+    /*
```
remove? (and below)

sherm1 commented on OpenSim/Simulation/Model/Geometry.cpp:
```diff
@@ -0,0 +1,216 @@
+void Mesh::createDecorativeGeometry(SimTK::Array_<SimTK::DecorativeGeometry>& decoGeoms) const
+{
+    const std::string& file = get_mesh_file();
+#if 0
```
remove?

sherm1 commented on OpenSim/Simulation/Model/Geometry.h:
```diff
@@ -0,0 +1,659 @@
+//+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+/* -------------------------------------------------------------------------- *
+*                              OpenSim:  Geometry.h                          *
+* -------------------------------------------------------------------------- *
```
Is this block repeated here?

sherm1 commented on OpenSim/Simulation/Model/Geometry.h:
```diff
@@ -0,0 +1,659 @@
+ * available to OpenSim component writers to express the Geometry of interest.
+ * The Geometry class handles serialization and also the translation to a set of
+ * DecorativeGeometry objects that gets passed to the Visualization system to be
+ * rendered. 
```
Optional style note: because of ubiquitous colorizing it doesn't really add anything now to begin every line with ` * `, and it eats up 3 characters out of our limited 80-character budget. It is also easier to reformat without dealing with the asterisks. So you can left justify block comments now after the opening `/**`.

sherm1 commented on OpenSim/Simulation/Model/Geometry.h:
```diff
@@ -0,0 +1,659 @@
+    virtual ~Geometry() {}
+    /** Interface methods to handle the Frame which the Geometry is attached to.
+    **/
+    /** setFrameName sets the name of the Frame of attachment **/
```
Shouldn't repeat the function name in the comment (and similar comments below).

sherm1 commented on OpenSim/Simulation/Model/Geometry.h:
```diff
@@ -0,0 +1,659 @@
+    //==========================================================================
+    // Compute Transform of this geometry relative to its base frame, utilizing 
+    // passed in state. Both transform and body_id are set in the passed-in 
+    // decorations as a side effect.
```
Is this an internal class? If not shouldn't these all be doxygen comments rather than `//` comments?

sherm1 commented on OpenSim/Simulation/Model/Model.cpp:
```diff
@@ -1012,7 +1017,15 @@ void Model::generateDecorations
-    Super::generateDecorations(fixed,hints,state,appendToThis);
+    ComponentList<Component> allComps = getComponentList();
+    ComponentList<Component>::const_iterator iter = allComps.begin();
+    iter++; // Skip Model itself
```
This seems dangerous to assume. Can you verify that this first component is actually the Model and blow up if not? I'm sure it is today but it seems unlikely that anyone in the future will realize that there is code that depends on that fact. Also if it isn't there already, there should be a comment where the component list gets created that says it is critical to put the Model first. Better would be to remove that assumption from this code if possible. It seems brittle to me.

sherm1 commented on OpenSim/Simulation/Model/ModelDisplay.h:
```diff
@@ -0,0 +1,87 @@
+ * National Institutes of Health (U54 GM072970, R24 HD065690) and by DARPA    *
+ * through the Warrior Web program.                                           *
+ *                                                                            *
+ * Copyright (c) 2005-2014 Stanford University and the Authors                *
```
2015?

sherm1 commented on OpenSim/Simulation/Model/ModelDisplayHints.h:
```diff
@@ -9,8 +9,8 @@
  *                                                                            *
- * Copyright (c) 2005-2012 Stanford University and the Authors                *
- * Author(s): Michael A. Sherman                                              *
+ * Copyright (c) 2005-2014 Stanford University and the Authors                *
```
2015?

sherm1 commented on OpenSim/Simulation/Model/ModelVisualizer.h:
```diff
@@ -181,9 +181,10 @@ class OSIMSIMULATION_API ModelVisualizer {
+                          const std::string&          geoFile,
                           bool&                       isAbsolute,
-                          SimTK::Array_<std::string>& attempts) const;
+                          SimTK::Array_<std::string>& attempts);
```
A method whose name begins with `find` should be `const`. Something is wrong if it has to be non-const.

sherm1 commented on OpenSim/Simulation/Model/PathPoint.cpp:
```diff
@@ -170,7 +168,7 @@ void PathPoint::updateGeometry()
     Transform position;
     position.setP(_location);
-    updDisplayer()->setTransform(position);
+    //updDisplayer()->setTransform(position);
```
remove?

sherm1 commented on OpenSim/Simulation/Model/PhysicalFrame.cpp:
```diff
@@ -117,11 +118,27 @@ void PhysicalFrame::scale(const SimTK::Vec3& aScaleFactors)
 
     SimTK::Vec3 oldScaleFactors;
-    getDisplayer()->getScaleFactors(oldScaleFactors);
+    //NewGeometry getDisplayer()->getScaleFactors(oldScaleFactors);
```
here and below can these commented-out lines be removed?

sherm1 commented on OpenSim/Simulation/Model/PointToPointSpring.cpp:
```diff
@@ -46,7 +46,7 @@ PointToPointSpring::PointToPointSpring()
     PointToPointSpring(const PhysicalFrame& body1, SimTK::Vec3 point1, 
                        const PhysicalFrame& body2, SimTK::Vec3 point2,
-                       double stiffness, double restlength )
+                                       double stiffness, double restlength )
```
formatting?

sherm1 commented on OpenSim/Simulation/Model/PointToPointSpring.cpp:
```diff
@@ -115,22 +115,12 @@ const PhysicalFrame& PointToPointSpring::getBody2() const
     const PhysicalFrame& body1 = getBody1();
     const PhysicalFrame& body2 = getBody2();
-
+    /**
```
remove commented-out code here and below?

sherm1 commented on OpenSim/Simulation/SimbodyEngine/Body.cpp:
```diff
@@ -79,8 +84,8 @@ void Body::constructProperties()
-    const SimTK::MassProperties& massProps = getMassProperties();
-    _internalRigidBody = SimTK::Body::Rigid(massProps);
+        const SimTK::MassProperties& massProps = getMassProperties();
+        _internalRigidBody = SimTK::Body::Rigid(massProps);
```
is the indent right here?

sherm1 commented on OpenSim/Simulation/SimbodyEngine/EllipsoidJoint.cpp:
```diff
@@ -202,6 +202,7 @@ void EllipsoidJoint::generateDecorations
     {
         // invoke parent class method
         Super::generateDecorations(fixed,hints,state,geometryArray); 
+        /*
```
remove (here and below)?

sherm1 commented on OpenSim/Tests/VisualizeModel/VisualizeModel.cpp:
```diff
@@ -0,0 +1,80 @@
+ * National Institutes of Health (U54 GM072970, R24 HD065690) and by DARPA    *
+ * through the Warrior Web program.                                           *
+ *                                                                            *
+ * Copyright (c) 2005-2014 Stanford University and the Authors                *
```
2015?

sherm1 commented:
Done reviewing -- looks good and was clearly a huge effort! I ran out of gas towards the end. General comment: there is lots of commented-out code. I'd like to see that removed rather than left behind, potentially to clutter up the code for eternity. If it is there as a reminder for upcoming work, that's fine but then it should be marked with a `TODO` and brief explanation of what it is there for.

Of everything there the one thing that bothered me most was the removal of `const` from a routine whose name was something like `findGeometryFile()`. I interpreted that as a dark hint that there is something ugly underneath.

aymanhab commented:
@sherm1 Great feedback as usual. The method findGeometryFile can be made const again but it's never used now since its only use is from Mesh::createDecorativeGeometry. That branch is commented out since it creates SimTK::PolygonalMesh rather than DecorativeMeshFile. The code was left behind since it needs to be moved to the visualizer process side (to follow the same logic) but I'm not sure if any other Simbody users are relying on it. We can discuss.

aseth1 commented on OpenSim/Common/Component.cpp:
```diff
@@ -180,7 +180,9 @@ void Component::connect(Component &root)
         connector.disconnect();
         try{
-            connector.findAndConnect(root);
+            const Component* connectTo = root.findComponent(connector.get_connected_to_name());
```
This was changed because findComponent is a weaker search than findAndConnect which also ensures the connectee is of the correct type and not purely a name match. If anything, `findComponent` should be templated to filter on Component type.

aseth1 commented on OpenSim/Simulation/Model/Appearance.h:
```diff
@@ -0,0 +1,153 @@
+    // PROPERTIES
+    //==========================================================================
+    /** @name Property declarations
+    These are the serializable properties associated with Appearance. **/
```
associated with BaseAppearance

aseth1 commented on OpenSim/Simulation/Model/Appearance.h:
```diff
@@ -0,0 +1,153 @@
+    OpenSim_DECLARE_PROPERTY(opacity, double,
+        "The opacity (0-1) used to display the object. ");
+    OpenSim_DECLARE_PROPERTY(representation, int,
+        "The representation (0:Hide, 1:Points, 2:Wire 2:Shaded) used to display the object. ");
```
Would be more informative  to associate the names of the enum rather than integer value. "Hide" should probably be "Hidden"

tkuchida replied:
I think you meant `2: Wire, 3:Shaded`.

aymanhab replied:
@aseth1 Good one, yes indeed.

aseth1 commented on OpenSim/Simulation/Model/Appearance.h:
```diff
@@ -0,0 +1,153 @@
+    //==============================================================================
+    /** @name Property declarations
+    These are the serializable properties associated with Appearance. **/
+    /**@{**/
```
associated with LineAppearance

aseth1 commented on OpenSim/Simulation/Model/Appearance.h:
```diff
@@ -0,0 +1,153 @@
+    /**@{**/
+    OpenSim_DECLARE_PROPERTY(thickness, double,
+        "The radius of the shape used to display the object. ");
+    /**@}**/
```
do you mean the thickness of the line? Where does radius come in for a line?

aseth1 commented on OpenSim/Simulation/Model/ModelComponent.h:
```diff
@@ -79,9 +78,18 @@ class ModelDisplayHints;
+/**@{**/
+public:
+    OpenSim_DECLARE_LIST_PROPERTY(geometry, Geometry,
+        "Geometry that can be attached to the Model Component");
```
Can be or is attached?

aseth1 commented on OpenSim/Simulation/Model/ModelComponent.h:
```diff
@@ -110,15 +118,20 @@ OpenSim_DECLARE_ABSTRACT_OBJECT(ModelComponent, Component);
-     * In case the ModelComponent has a visual representation (VisualObject), override this method  
-     * to update it. This is typically done by recomputing anchor points and positions based 
-     * on transforms obtained from current state.
+     * getGeometrySize returns the number of published Geometry objects 
```
"Get the number of Geometry components attached to this ModelComponent"?

aseth1 commented on OpenSim/Simulation/Model/ModelComponent.h:
```diff
@@ -110,15 +118,20 @@ OpenSim_DECLARE_ABSTRACT_OBJECT(ModelComponent, Component);
      */
-    virtual void updateDisplayer(const SimTK::State& s) const {};
-
+    int getGeometrySize() const {
```
`getNumGeometry()` perhaps?

aseth1 commented on OpenSim/Simulation/Model/AppearanceMap.h:
```diff
@@ -0,0 +1,100 @@
+//=============================================================================
+/**
+ * A class that holds the map between object PathName and Appearance
+ *
```
Who uses an appearance map? Is it a utility class to serve some other object and not intended for public use?

aymanhab replied:
@aseth1 Yes, it's unused now. At some point we considered having a mask applied on top of the Appearance specified in the model but as of now it's not wired. I'd be ok with removing it until we have a use case.

aseth1 commented on OpenSim/Simulation/Model/AppearanceMap.h:
```diff
@@ -0,0 +1,100 @@
+    OpenSim_DECLARE_CONCRETE_OBJECT(GeometryAppearance, Object);
+public:
+    OpenSim_DECLARE_PROPERTY(geometryID, std::string,
+        "PathName for the geometry.");
```
What is a `geometryID` and what is a `PathName`? Why two names?

aseth1 commented on OpenSim/Simulation/Model/AppearanceMap.h:
```diff
@@ -0,0 +1,100 @@
+    /**@{**/
+    OpenSim_DECLARE_LIST_PROPERTY(AppearanceList, GeometryAppearance,
+        "Map between Geometry objects and user specified display preferences");
+    /**@}**/
```
What are "display preferences"?
Is this A list of Geometry and their associated Appearances for visualization in the GUI ?

aymanhab replied:
@aseth1 Yes that's the mask that can override the colors in the model. unused for now.

aseth1 commented on OpenSim/Simulation/Model/ExpressionBasedBushingForce.cpp:
```diff
@@ -595,39 +595,6 @@ getRecordValues(const SimTK::State& state) const
 /**
  * Get the visible object used to represent the spring.
  */
-VisibleObject* ExpressionBasedBushingForce::getDisplayer() const
```
So awesome to see these "Displayer" methods go away! Yay!

aseth1 commented on OpenSim/Simulation/Model/ModelDisplay.h:
```diff
@@ -0,0 +1,87 @@
+
+namespace OpenSim {
+
+class Body;
```
Why are these forward declared?

aseth1 commented on Applications/IK/test/constraintTest_ikOnly.osim:
```diff
@@ -5941,7 +5941,7 @@
 				</Body>
-				<Body name="toes_l">
+				<Body name="toes_left">
 					<mass>       0.09750000 </mass>
```
Why was this renamed? Not that the naming is bad, but the other bodies were not renamed for clarity, so I assume this was renamed for some other reason.

aymanhab replied:
@aseth1 Indeed, the model has a Body and Joint with the same name so findComponent was returning the wrong type and throwing an exception that it's not finding the PhysicalFrame/Body. If the search is done by name then we need to make the name unique and hopefully enforce the uniqueness downstream.

aseth1 replied:
That is why I had switched the procedure of making a connection to use `Connector<C>::findAndConnect()` which does filter on Component type. I think we should understand why that wasn't working for you in another instance, because it was certainly working in this case.

aseth1 commented on OpenSim/Simulation/Model/PhysicalFrame.h:
```diff
@@ -140,17 +136,16 @@ class OSIMSIMULATION_API PhysicalFrame : public Frame {
+    /** Add a Mesh specified by file name to the list of Geometry owned by the ModelComponent.
+    Scale defaults to 1.0 but can be changed on the call line.
+    */
+    void addMeshGeometry(const std::string &aGeometryFileName, const SimTK::Vec3 scale = SimTK::Vec3(1));
```
Why the special case for adding Mesh geometry and why only on PhysicalFrame? If MeshGeometry is Geometry, then `ModelComponent::addGeometery()` should be sufficient. Why not create the Geometry first then add it like any other Geometry?

aymanhab replied:
@aseth1 Good point, it's mainly for convenience as used by our API examples and test cases, but these could change to create the Mesh instead. Why on Frames? because in these cases the user doesn't need to explicitly state the Frame that the Geometry is attached to, so more convenience and less error prone.

aseth1 replied:
That's a nice service (to use itself as the Frame of the Geometry) that a Frame in general could provide and not only PhysicalFrame. Wouldn't you want that capability for adding any Geometry and not just MeshGeometry? We could provide a virtual `extendAddGeometry()` method that would allow derived Component to do more handling of Geometry that is being added to it. In which case, we should use `Component::addGeometry(const Geometry& geom)` and make a copy especially if we intend to modify it. It would then allow Geometry of one Component (say a Frame) to be  added to another Frame and have its frame connector updated accordingly. It does assume then that the Component owns the Geometry, which I think it does by virtue of `<geometry>` being a list property of the Component.

aseth1 commented on OpenSim/Simulation/Model/PathPoint.h:
```diff
@@ -73,16 +74,11 @@ OpenSim_DECLARE_CONCRETE_OBJECT(PathPoint, Object);
-    VisibleObject _displayer;
 
-    /* const*/ PhysicalFrame* _body; // Not const anymore since the body's displayer is not const
+    const PhysicalFrame* _body; // Not const anymore since the body's displayer is not const
```
Comment is no longer relevant.

aseth1 commented on OpenSim/Simulation/Model/Geometry.h:
```diff
@@ -0,0 +1,627 @@
+    //--------------------------------------------------------------------------
+public:
+    /// Default constructor, does nothing
+    Geometry()
```
A convenience constructor that takes the Frame that the Geometry should be connected too would be helpful and eliminate the clutter in the examples. See Joint constructors that take parent and child PhysicalFrames.

aseth1 commented on OpenSim/Simulation/Model/Geometry.h:
```diff
@@ -0,0 +1,627 @@
+    /** Return a reference to the actual Frame to which this Geometry
+    is attached. */
+    const PhysicalFrame& getFrame() const;
+    //==========================================================================
```
Setting the Frame directly when you have it in hand would be helpful too!

aseth1 commented on OpenSim/Simulation/Model/Geometry.h:
```diff
@@ -0,0 +1,627 @@
+ * LineGeometry is a utility class used to abstract a line segment.
+ * It is used by muscle segments so that it's as small and useful as possible.
+ */
+class OSIMSIMULATION_API LineGeometry : public Geometry
```
This class does not abstract a line segment. A true line would have Points that could live in different Frames. This is really a fixed length and direction ray with respect to a Frame.

If truly used to draw muscles, it should depend on Points and not Vec3s. How is this used to draw a muscle segment across two bodies?

aymanhab replied:
@aseth1 We have used this object as is for a while now to display muscles in the API visualizer. For lines that cross/connect multiple bodies, everything is drawn in ground Frame.

aseth1 commented on OpenSim/Simulation/Model/ModelDisplay.h:
```diff
@@ -0,0 +1,87 @@
+ * but in the future can be expanded to include search paths for Geometry, Texture
+ * Renderer preferences, lights, cameras etc.
+ *
+ * @author Ayman Habib
```
I likely do not adhere to this and I should, but we should strive to describe new classes using existing abstractions and describe new concepts/abstractions introduced by the class or a logical grouping of classes. For example, what are Display Attributes? Why is a group of these in a new class?
Is a "full Model" different from a partial Model or even a ModelComponent. If so, why? If we need a special Component that manages the Appearances of Components in a Visualizer why is it a) called ModelDisplay and b) an Object and not a Component itself so it can leverage system resources (like realization Stages) to determine what can be visualized and how. E.g no forces when realized only to position (from. IK).

aymanhab commented:
@aseth1 @sherm1 Based on discussion in dev meeting remaining issues to be resolved before merging are:
- Add //TODO before commented out code where appropriate
- Fix iterator use case that skips root model either by modifying client code or iterator itself.
- Add constructor to Geometry that takes a Frame input.
- Resolve whether it's possible to restore findAndConnect. Since @aseth1 suggested a different place to populate the tree iterator, I'll provide a simple example from the test cases and he'll suggest solution(s).
- Possible renaming of ModelDisplayHints to VisualizerHints and reduce contents.

aseth1 commented on OpenSim/Simulation/Model/Appearance.h:
```diff
@@ -0,0 +1,151 @@
+ * Geometry that have a surface so that it can be textured can use the subclass
+ * Appearance, while schematic line drawings (e.g. Arrows, Frames) can use 
+ * LineAppearance which offers thickness. 
+ *
```
What is an Appearance? I think we stick to Visualizer (as the abstraction that visualizes Geometry), visualize (the action) and visualization as the noun for the product of a Visualizer then we can get rid of the use of Displayer, display/ed and the display (noun).

Still a fuzzy issue is what is the difference between an Appearance and a Property of a piece of Geometry? When should an attribute be part of the Appearance? For example, thickness. How thick a line could be a Property of a Line itself? Then there would be no need for LineAppearance, would there? Being as precise as we can about what is (and therefore what is not) an Appearance would help a lot. For example, If I create a new Geometry called BlobGeometry and it has a <lumpy> attribute which defines its Geometry and its visualization (similar to a sphere's radius) should I create a LumpyAppearance or BlobAppearance as well or instead? I don't know the answer.

aseth1 commented on OpenSim/Simulation/Model/Appearance.h:
```diff
@@ -0,0 +1,151 @@
+    /**@{**/
+    OpenSim_DECLARE_PROPERTY(thickness, double,
+        "The radius of the tube used to display the cooresponding lines. ");
+    /**@}**/
```
How about just "The thickness used to visualize a LineGeometry" ?

aymanhab commented:
Here we go. From the last summary list I implemented the first three items, 4th item I believe is working as expected, @aseth1 thinks otherwise and will look into restoring the use of findAndConnect in Component.cpp (commented out now) instead of the previous version currently in code. testFrames shows an example for why this fails on a simple double_pendulum example. For item 5, I changed the name of the class ModelDisplay into ModelVisualPreferences which is a reasonable name that relates to "Visualization",  the contents will evolve as we do more  Visualization since it's self contained  and this PR is fundamentally about Geometry rather than rendering. Ready for final pass thru or merge.

sherm1 commented:
In the interest of moving forward, I think it would be good to merge this asap and do follow-ups with issues and separate PRs. It is too big to review and re-review.

aseth1 commented:
@aymanhab if you did not fix the `findAndConnect()` issue then it must mean that the testIK case with "constraintTest_ikOnly.osim" remains weakened (e.g. hack to rename toes_l to toes_left). I am reluctant to merge without this being restored and the renaming undone. The issue as we discussed is that the iterator tree must be built and searchable prior to the individual components being connected. As long as that happens there shouldn't be a problem. I find it odd that it worked on the case which is now broken (in constraintTest_ikOnly.osim and required the rename) and all other connections. Are you sure you are giving the `findAndConnect()` the correct root to search through? Just a thought.

aseth1 commented:
@sherm1 @aymanhab 
1. We shouldn't introduce obvious new bugs (the rename of tag in in a test file is a clear red flag! It will come back to bite us in a big way!). 
2.  The list of Todo's/cleanup should be well defined and agreed upon as an open issue.

Otherwise, I agree that the PR is getting unwieldy.

aseth1 commented on OpenSim/Simulation/Model/Geometry.h:
```diff
@@ -0,0 +1,641 @@
+        "Default appearance for this Geometry");
+
+    enum DisplayPreference {
+        Hide = 0,       ///<Hide geometry from display
```
I am confused. Why is the DisplayPreference not a part of Appearance? For me the Draw prefix adds little. `Hidden`, `Points`, `WireFrame`, `Surface` seem clear enough. Why is an explicit `Default` required is that not implied if one is not specified?

chrisdembia commented:
I don't like the idea of merging code into master that hasn't been reviewed to everyone's satisfaction. I don't think the workflow of merge-now-and-file-new-issues is really safe, because nothing requires us to resolve the new issues. 

I would really really prefer to see a system of long term feature branches that are reviewed in small chunks. 

For this PR, I would create a new empty feature branch geometry-review. @aymanhab can open a PR of a small subset of his geometry_v2 changes to merge into geometry-review. When that PR is accepted, @aymanhab can create another bite-size PR. I know this process **sounds** slow, but I believe it will be much faster--and will avoid the bugs--than reviewing a large PR like this.

chrisdembia commented:
Let me say a little more...we have to decide on an official policy for PRs/features that introduce large-scale changes. I don't think merge-now-and-file-new-issues should be the official policy. We can establish the new policy before or after this PR. But if we agree that we need a new policy, then we should avoid the debt as soon as possible, starting with this PR. I really think we'll regret it otherwise.

aymanhab commented:
@chrisdembia Thanks for pitching in your opinion. In general my worry about feature branches is that they feel like they just kick the can down the road rather than solve the problem in these situations where there are major code changes across the whole system as was the case in utilizing Frames, or with this new Geometry PR. There was great feedback along the way and it has been incorporated into the PR already. The one single issue under discussion here is a change that we can't agree if it's a bug or use of a function prematurely (issue #418). I provided an example to show that the issue is there in master already and we can resolve it. 

Historically, over our use of github, all large PRs have undergone many revisions and it became unwieldy at the end so it feels that there's a need for a new policy indeed.

sherm1 commented:
I've been assuming that if these big PRs could have been done in smaller pieces they would have been. Certainly a good policy (which I think we already have) is to favor small PRs over big ones when possible. I'm not sure how having a separate branch improves the situation -- can you explain more, Chris?

aseth1 commented:
I have share the sentiment with Chris. It is virtually impossible to review the design and the abstractions in themselves. A feature branch with reviewed abstractions and examples would have been much easier to review and would isolate from the massive changes due to a merge. Now they are all mangled and it is very hard to navigate the logic through GitHub. We should discuss it.

chrisdembia commented:
> I've been assuming that if these big PRs could have been done in smaller pieces they would have been.

Hmm.. good point. I hadn't thought of that. Ajay, Matt and I were talking about how perhaps we are still working in the culture of SVN, where large sweeping changes like frames, geometry were the only way to operate, given the high cost of branching. Maybe it's idealistic of me to think that some of these large features could occur through smaller PR's. 

> Can you explain more, Chris?

First, only completed features should be merged into master. However, completing some features (frames, geometry) requires more code changes than are reasonable for people to review at once. So I want to decouple merges into master from the amount of code that gets reviewed in one PR. That means introducing a new type of PR in which it **is** okay to merge incomplete features. That is, merges into long term feature branches. I understand that to have frames or geometry working, some 100 files have to change. But that doesn't mean all 100 files have to be reviewed at once. I don't think we're using branches the best way we can.

In fact, I've been practicing this feature branch workflow myself, first with CMA in simbody, and also with unix-compatibility in OpenSim (https://github.com/opensim-org/opensim-core/issues/195).

1. I create an empty feature branch, e.g., cmaes.
2. I write a minimally functional implementation in a separate branch,e.g. cmaes-first, and open a PR to merge cmaes-first into cmaes.
3. While cmaes-first is being reviewed, I create a new branch off of cmaes-first, called cmaes-second or something like that. I continue working on the next part of the feature while cmaes-first is being reviewed.
4. etc.

This workflow is more complicated for large API changes. Perhaps we can discuss more in the dev meeting.

chrisdembia commented:
Perhaps if a PR is going to introduce a new class, the header file for the class should be reviewed first, without any implementation as a rule. This means that for reviewers to understand the class, the interface must be documented to sufficient clarity. The first PR or first few PR's can introduce one or all of the new interfaces for the feature, depending on how necessary it is to have all the interfaces to understand how they will work together. I also strongly believe in test driven development, so I think the first PRs should be accompanied by tests (that wouldn't even compile yet) that show how the proposed interface would be used.

@aymanhab I shouldn't have spoken about this PR in particular. I haven't been following it so I don't know its status. My apologies.

chrisdembia commented:
Also, I think a contributor is likely to be much less frustrated when a reviewer tells them to change their interface if they haven't yet started implementing it.

sherm1 commented:
> I think a contributor is likely to be much less frustrated when a reviewer tells them to change their interface if they haven't yet started implementing it

That sounds good for introducing new functionality with new classes. I'm not sure that describes our problematic PRs though. The big ones (like this one) involve massive changes in implementation throughout the code, and most of the discussion seems to revolve around those changes. New interfaces are a very small part of the work.

chrisdembia commented:
> most of the discussion seems to revolve around those changes.

hmm good point...

aymanhab commented:
@aseth1 With tag renaming undone and issue #418 fixed and merged into this branch, please let me know if you'll be reviewing more or if we should merge as I'm not aware of any outstanding issues. This will also address issue #419. Thanks.

aseth1 commented:
@aymanhab I am still confused between Geometry and Appearance. I don't understand why there is a LineAppearance for example and not a LineGeometry. Why is thickness part of appearance and not the Line itself? Universal appearance attributes like visibility, transparency, and draw mode make sense to me as separate from the Geometry but not thickness or radii.  Also confusing is the difference between Appearance and DisplayPreference. If they are the same thing then I would like DisplayPreference to be removed from all comments and descriptions.

aymanhab commented:
Thanks @aseth1 for the feedback and for clearing up what you find confusing. This PR is mainly about Geometry, the inclusion of Appearance is mainly to test the wiring (using the current GUI) but will evolve with the capabilities of the Visualizer itself. That said, here's the answer to your questions:
- Geometry is the pure shape (Sphere, Cone, Mesh, ...) that has intrinsic attributes.
- Appearance is how we want the shape rendered (what color, DisplayPreference=representation: smooth-shaded/flat-shaded/none etc.) eventually will include shininess, etc.

Depending on the shape of the Geometry some "Appearance" attributes may not apply, for example texture makes sense on a surface/manifold but not for a Line or a Frame, which would have thickness, maybe pattern instead or Text which may have a Font, billboarding etc. and that's what the different classes under Appearance are for. 

DisplayPreference is part of Appearance and is an enum that we use to indicate the representation as we've done in the past .The class name itself is internal, and the methods to set/get-it are called set/getRepresentation().

Please let me know if that explains or if you have any further questions.

aseth1 commented:
>•	Geometry is the pure shape (Sphere, Cone, Mesh, ...) that has intrinsic attributes.
•	Appearance is how we want the shape rendered (what color, DisplayPreference=representation: smooth-shaded/flat-shaded/none etc.) eventually will include shininess, etc.

These definitions are not precise enough. For example, what is a Point or a Line? Are they Geometry? Do I need a piece of Geometry to visualize it? If I choose a Sphere to represent a point (like we do for Marker) then is its radius an Appearance? Radius has nothing to do with the intrinsic attributes of a Point. But in another context we have a Sphere as a contact element then its radius has everything to do with its physics. How should we visualize Points in OpenSim? Should there be `PointGeometry` like there is `FrameGeometry` and `LineGeometry`? Seems like these are just `Sphere` and `Cylinder` under another name so they can have `PointAppearance` and `LineAppearance` to change otherwise "intrinsic attributes" of the Geometry. I am not a fan of these `FrameGeometry`, `LineGeometry`, etc.. classes. We have a Frame class and it can draw Geometry, so why do we need `FrameGeometry` as another class to visualize it? Why couldn’t `Frame::generateDecorations()` be used to draw its own axes?

As another example, for `AnalyticalGeometry` how will we control how many facets are used to visualize the `Sphere` or `Ellispoid`? An `AnalyticalAppearance`? Do you envision `<ComponentType>Geometry` for every Component that has Geometry as well as parallel `<ComponentType>Appearance`? Is this really scalable/sustainable?

I really like the idea of having an Appearance that only changes the visualization and not the physics of the model. The current approach is coming at the expense of duplicate code and multiple (practically empty) classes (e.g. to expose one property). 

Would it be easier/cleaner to then to have a new AppearanceProperty of an Object/Component? It seems the whole purpose of Appearance is to expose properties that can affect the appearance of a Component and not its physics.  None of the Appearance classes do anything other than contain properties (as far as I can tell) and that makes me question their existence. Unlike regular properties, AppearanceProperties can have their values changed and they would not mark the Object as not being up-to-date (thus would not trigger setting up from properties, connecting, etc.. to re-initSystem). That is my main worry for including things that only affect visualization as properties. We must have a clear separation between physics (properties) and visuals (appearance) and the latter should have absolutely no effect on the former.

Perhaps another way of controlling appearance without polluting the modeling/physics API with Appearances or AppearanceProperties is for the Visualizer to employ a Visitor to apply appearances based on the Component (and its Geometry) that are encountered.

Besides the big picture above here is a summary of issues:

1. The `BaseAppearance` class should just be the `Appearance`.
2. Appearance should define Representation (currently still DisplayPreference) not Geometry. 
3. The current Appearance class should be SurfaceAppearance since textures can only be applied to surfaces.
4. Geometry does not derive from ModelComponent so it should remain in Common and should not have been moved to Simulation/Model
5. DisplayPreference Enum should be renamed Representation since properties and accessors refer to “representation”. All use of DisplayPrefence (including comments) should be eliminated.
6. Eliminate “Draw” from the enumerated types of Representation.  The Draw prefix adds little.  “Hidden’, “Vertices”, “WireFrame”,  “Surface” seem clear enough. Why is an explicit Default required is that not implied if one is not specified?
7. The Appearance::Representation name (string) should be serialized (e.g. Hidden, Vertices, etc…) (as in the case of Coordinate::MotionType) opposed to its integer value.
8. Why LineGeomerty class vs. a Line? Would it be generally useful to have a Line class that connects two Points? And these points can live on different Frames? Again, why can’t Line::generateDecorations() not be employed to draw a line instead of a LineGeometry and a LineAppearance?

sherm1 commented:
Ajay, I hope you are feeling better! You make many good points above, but IMO this is too much to put on this one PR. I think we should move those to a new 4.0 issue (or maybe several 4.0 issues) and merge this one as-is. Because it involves so many files (100+) it is holding up other PRs now. It has been pending for a month and a half. We must move forward ...

aymanhab commented:
Thanks @aseth1 for the great feedback. 
Issue #480 has been opened to start the discussion regarding Appearance which will be refined as we get to display Text, Surfaces and icon-like objects. The parts of the discussion above related to Appearance has been copied for future reference.

As to the other issues:
A Line can definitely create a rubberband line that lives in two frames, the new Geometry presented in this PR is only the Geometry that can be serialized in XML. Component developers can do whatever they want as long as the visualizer can support it. A point is a location, it can use any "Geometry" to render itself, same as Vector that can employ Arrow, curvedArrow, or any shape it desires to display itself.

aseth1 commented:
@aymanhab @sherm1 I apologize for the long over-due review.  I wanted to speak in person but my cold has held me back. 
I feel we cannot ignore the issues so I wanted to get them down thus the lengthy comment. Thanks Ayman for opening a related issue. I agree that the point list of issues could be handled by another PR. 

The bigger question is whether or not the Geometry + Appearance pairing is the right approach for separating physical geometry from appearance properties since most Components need both. My main concern is how it will scale.  If the consensus is to merge it in and figure it out soon after, we should have an issue open so we don't forget.
I also restarted the travis build.

aymanhab commented:
Awesome, thanks much Ajay, I sure value your thoughtful feedback and will
make sure nothing falls thru the cracks. All the best.
On Jun 8, 2015 4:55 PM, "Ajay Seth" <notifications@github.com> wrote:

> @aymanhab <https://github.com/aymanhab> @sherm1
> <https://github.com/sherm1> I apologize for the long over-due review. I
> wanted to speak in person but my cold has held me back.
> I feel we cannot ignore the issues so I wanted to get them down thus the
> lengthy comment. Thanks Ayman for opening a related issue. I agree that the
> point list of issues could be handled by another PR.
>
> The bigger question is whether or not the Geometry + Appearance pairing is
> the right approach for separating physical geometry from appearance
> properties since most Components need both. My main concern is how it will
> scale. If the consensus is to merge it in and figure it out soon after, we
> should have an issue open so we don't forget.
> I also restarted the travis build.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opensim-org/opensim-core/pull/417#issuecomment-110177024>
> .
>

sherm1 commented:
Ajay, can you verify that issue #480 is sufficient and if anything is missing add it there or to other issues? Then you can merge it (or delegate it)!

aymanhab commented:
Also refer to issue #482

aseth1 commented:
@aymanhab congrats for getting such a big addition in! Looking forward to the updates.

aymanhab commented:
Great, thanks much @aseth1 and I sure look forward to more interactions as we move forward. Thanks all for the reviews and great feedback throughout :fireworks:

jenhicks commented:
Woo hoo! Thanks guys.

@aymanhab One final thing that is missing is notes for the changelog. Can you summarize in a comment or edit the changelog directly? Thanks!

sherm1 commented:
Awesome, congratulations Ayman and reviewers! 

Ayman, I see the branch is still there -- please delete if appropriate to avoid branch explosion.

chrisdembia commented:
"branch :boom:"
